### PR TITLE
Feature not equal

### PR DESCRIPTION
**Pull request by pmqtt**
Hello folks,
I have implement != operator
I hope it is useful !

### CODE REVIEW

File: explorer/data/prelude.carbon [hunk 1]
```diff
@@ -156,6 +156,10 @@ interface GreaterEqWith(U:! Type) {
   fn GreaterEq[me: Self](other: U) -> bool;
 }
 
+interface NotEqWith(U:! Type){
+  fn NotEq[me: Self](other: U) -> bool;
```
Comment (geoffromer) [inline explorer/data/prelude.carbon]: This doesn't match the [current design](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/expressions/comparison_operators.md), where the `NotEqual` method is part of the same interface as `Equal`.

File: explorer/data/prelude.carbon [hunk 2]
```diff
@@ -264,6 +268,30 @@ impl String as GreaterEqWith(Self) {
...
 
+impl i32 as NotEqWith(Self) {
+  fn NotEq[me: Self](other: Self) -> bool {
+    var comp: Ordering = me.(CompareWith(i32).Compare)(other);
+    match (comp) {
+      case Ordering.Equivalent() => {
+        return false;
+      }
+    }
+    return true;
```
Comment (geoffromer) [inline explorer/data/prelude.carbon]: This seems unnecessarily verbose, and also a little surprising: if we define `!=` in terms of another operation, `==` seems like a more natural choice than `CompareWith`.

```suggestion
  return not __intrinsic_int_eq(me, other);
```

File: explorer/data/prelude.carbon [hunk 3]
```diff
@@ -264,6 +268,30 @@ impl String as GreaterEqWith(Self) {
...
+impl String as NotEqWith(Self) {
+  fn NotEq[me: Self](other: Self) -> bool {
+    var comp: Ordering = me.(CompareWith(String).Compare)(other);
+    match (comp) {
+      case Ordering.Equivalent() => {
+        return false;
+      }
+    }
+    return true;
+  }
```
Comment (geoffromer) [inline explorer/data/prelude.carbon]:
```suggestion
    return not __intrinsic_str_eq(me, other);
```

File: explorer/data/prelude.carbon (hunk 2 above)
Comment (pmqtt) [reply explorer/data/prelude.carbon]: Changed!

File: explorer/data/prelude.carbon (hunk 3 above)
Comment (pmqtt) [reply explorer/data/prelude.carbon]: Changed!

File: explorer/data/prelude.carbon (hunk 1 above)
Comment (jonmeow) [reply explorer/data/prelude.carbon]: I'd agreed to doing this for the various comparisons, which by design, belong on an OrderedWith interface. The problem is that `default fn` isn't implemented, but people do want to use the features. I think the incremental cost of accepting changes such as this one is pretty low, and we can clean up later, so my own opinion is to accept things like this with a TODO indicating the necessary cleanup.

Comment (pmqtt) [reply explorer/data/prelude.carbon]: Sorry... I had the design paper not longer in my mind. You are complety right that is  unnecessarily verbose.

Comment (geoffromer) [reply explorer/data/prelude.carbon]: > I'd agreed to doing this for the various comparisons, which by design, belong on an OrderedWith interface. The problem is that `default fn` isn't implemented, but people do want to use the features.

I see. In that case I'm OK either way, but I prefer the way @pmqtt has now done it, which matches the design but with `default` omitted. Are you OK with this as it is?

File: explorer/testdata/mixin/fail-circular-mixing.carbon [hunk 4]
```diff
@@ -13,7 +13,7 @@ __mixin M1 {
   fn F1[me: Self](x: Self) -> Self{
      return x;
   }
-// CHECK: COMPILATION ERROR: {{.*}}/explorer/testdata/mixin/fail-circular-mixing.carbon:[[@LINE+1]]: 'M3' has not been declared yet
+  // CHECK: COMPILATION ERROR: {{.*}}/explorer/testdata/mixin/fail-circular-mixing.carbon:[[@LINE+1]]: 'M3' has not been declared yet
```
Comment (geoffromer) [inline explorer/testdata/mixin/fail-circular-mixing.carbon]: Can you revert these unrelated changes?

File: explorer/interpreter/type_checker.cpp [hunk 5]
```diff
@@ -2123,6 +2123,20 @@ auto TypeChecker::TypeCheckExp(Nonnull<Expression*> e,
...
+              BuiltinMethodCall{"NotEqual", op.arguments()[1]});
+          if (!converted.ok()) {
+            // We couldn't find a matching `impl`.
+            return CompilationError(e->source_loc())
+                   << *ts[0] << " is not equal comparable with " << *ts[1]
+                   << " (" << converted.error().message() << ")";
+          }
+          op.set_rewritten_form(*converted);
+          return Success();
+        }
```
Comment (zygoloid) [inline explorer/interpreter/type_checker.cpp]: Can you reduce the duplication here? The only difference between these two cases is (should be) the `"Equal"` versus `"NotEqual"`, so perhaps we could have a single implementation with an `op.op() == Operator::Eq ? "Equal" : "NotEqual"`, or perhaps the implementation could be factored out into a lambda to which the right string is passed?

**Review by zygoloid — COMMENTED** [review]
Generally-speaking, this looks good. It makes sense to me to require a `NotEqual` implementation until `default fn` is implemented.

File: explorer/interpreter/type_checker.cpp (hunk 5 above)
Comment (pmqtt) [reply explorer/interpreter/type_checker.cpp]: Sure

Comment (pmqtt) [reply explorer/interpreter/type_checker.cpp]: DONE!

File: explorer/testdata/mixin/fail-circular-mixing.carbon (hunk 4 above)
Comment (pmqtt) [reply explorer/testdata/mixin/fail-circular-mixing.carbon]: DONE!

**Review by geoffromer — COMMENTED** [review]
Looks good to me. Leaving approval to @zygoloid

**Review by zygoloid — APPROVED** [review]
Looks great, thanks!
